cairo/src/utils/dict.cairo

```diff
@@ -136,3 +137,83 @@ func hashdict_write{poseidon_ptr: PoseidonBuiltin*, dict_ptr: DictAccess*}(
     let dict_ptr = dict_ptr + DictAccess.SIZE;
     return ();
 }
+
+func hashdict_finalize{range_check_ptr}(
+    dict_ptr_start: DictAccess*,
+    dict_ptr: DictAccess*,
+    original_mapping_start: DictAccess*,
+    original_mapping_end: DictAccess*,
+    should_merge: felt,
+) -> (DictAccess*, DictAccess*) {
+    alloc_locals;
+
+    if (should_merge == FALSE) {
+        let (squashed_dict_start: DictAccess*) = alloc();
+        let (squashed_dict_end) = squash_dict(dict_ptr_start, dict_ptr, squashed_dict_start);
+        let (prev_values_start, prev_values_end) = prev_values(
+            squashed_dict_start, squashed_dict_end
+        );
+        tempvar dict_ptr_start = prev_values_start;
+        tempvar dict_ptr = prev_values_end;
+        tempvar range_check_ptr = range_check_ptr;
+    } else {
+        tempvar dict_ptr_start = dict_ptr_start;
+        tempvar dict_ptr = dict_ptr;
+        tempvar range_check_ptr = range_check_ptr;
+    }
+    let dict_ptr_start = cast([ap - 3], DictAccess*);
+    let dict_ptr = cast([ap - 2], DictAccess*);
+    let range_check_ptr = [ap - 1];
+
+    if (cast(original_mapping_end, felt) == 0) {
+        // No parent mapping, just return the current dict.
+        return (dict_ptr_start, dict_ptr);
+    }
+
+    let segment_to_merge = dict_ptr_start;
+    let len_to_merge = dict_ptr - dict_ptr_start;
+    let dst_ptr = original_mapping_end;
+    memcpy(dst_ptr, segment_to_merge, len_to_merge);
+    let new_original_mapping_end = dst_ptr + len_to_merge;
+    return (original_mapping_start, new_original_mapping_end);
+}
+
+// @notice Given a dict segment (start and end), returns a new dict segment with (key, prev_value, prev_value) for each key.
+// @dev Expectes the given dict to be squashed, with one DictAccess instance per key.
+// @param dict_ptr_start: The start of the dict segment.
+// @param dict_ptr_stop: The end of the dict segment.
+// @return prev_values_start: The start of the new dict segment.
+// @return prev_values_end: The end of the new dict segment.
+func prev_values{range_check_ptr}(dict_ptr_start: DictAccess*, dict_ptr_stop: DictAccess*) -> (
+    prev_values_start: DictAccess*, prev_values_end: DictAccess*
+) {
+    alloc_locals;
+    let (local prev_values_start: DictAccess*) = alloc();
+
+    tempvar prev_values = prev_values_start;
+    tempvar dict_ptr = dict_ptr_start;
+
+    loop:
+    let prev_values = cast([ap - 2], DictAccess*);
+    let dict_ptr = cast([ap - 1], DictAccess*);
+
+    let is_done = is_zero(dict_ptr_stop - dict_ptr);
+
+    tempvar prev_values_end = prev_values;
+    jmp end if is_done != 0;
```

Comment:
this will do one useless `tempvar` per loop, do an early return instead if there is nothing to copy, and check the condition at the end. You can also do the control flow in a hint, and just assert outside of the `loop` that `dict_ptr = dict_ptr_stop`